C++ API parity: at::Tensor::detach

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26568 C++ API parity: at::Tensor::register_hook
* #26332 C++ API parity: at::Tensor::requires_grad_
* **#26251 C++ API parity: at::Tensor::detach**

Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

pbelevich committed:
C++ API parity: at::Tensor::detach

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

yf225 commented on test/cpp/api/tensor.cpp:
```diff
@@ -406,3 +406,31 @@ TEST(TensorTest, Requires_grad_) {
+  auto y = x * x;
+  const auto y_detached = y.detach();
+  ASSERT_FALSE(y.is_leaf());
+  ASSERT_TRUE(y_detached.is_leaf());
```
We might want to check `y_detached.requires_grad() == false`, based on https://pytorch.org/docs/stable/tensors.html#torch.Tensor.detach `The result will never require gradient`.

yf225 commented on test/cpp/api/tensor.cpp:
```diff
@@ -406,3 +406,31 @@ TEST(TensorTest, Requires_grad_) {
+  ASSERT_THROWS_WITH(y.detach(), message);
+}
+
+TEST(TensorTest, Detach_) {
```
I feel that we might want to name this `DetachInplace`, to preserve the naming convention (of not using underscore). Same for other tests as well.

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

yf225 reviewed (approved):
Thanks a lot for the thorough tests @pbelevich !

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::detach"


Differential Revision: [D17427578](https://our.internmc.facebook.com/intern/diff/D17427578)

[ghstack-poisoned]
